### ISSUE TITLE
Fixed Width SearchBox is broken on Insiders Edge

### ISSUE THREAD

**Issue by hoovercj** [OfficeDev/office-ui-fabric-react#6417]
### Please provide a reproduction of the bug in a codepen:
 
Website: https://developer.microsoft.com/en-us/fabric#/components/searchbox
Codepen: https://codepen.io/anon/pen/LJMLby

### Bug Report 
- __Package version(s)__: All
- __Browser and OS versions__: Windows Insiders (17754.1) and Edge 44.17754.1.0
  
#### Describe the issue:
 
The example "SearchBox with fixed width" from the Office UI Fabric website renders incorrectly in the insiders version of Edge. It works in IE, Chrome, Firefox, Safari, and the "release" version of Edge (Spring 2018 release).

In the above links, put focus into the narrow "Search" box, type content, and notice how the "X" icon begins to be pushed outside the search box. Now move focus out of the search box so the "Search icon" appears again. Notice how the "X" icon is pushed even further outside the search box.

Contrast that with the other search boxes whose "X" icon appears inside the search box.

I have opened [a bug](https://userfeedback.visualstudio.com/UIF/_workitems/edit/7367358) against Edge for this, but I thought y'all would like to be aware of it, too.

**Comment by micahgodbolt** [comment]
The 'input' is set to flex: 1 1 0px. If that flex shrink value was getting lost, that would cause this issue.

Can you try this pen to see if explicit flex-shrink helps?

https://codepen.io/micahgodbolt/pen/MqZVBa

**Comment by hoovercj** [comment]
That didn't help, no.

**Comment by micahgodbolt** [comment]
Okay, worth a stab. From your description it sounds like the input isn't respecting the flex-shrink value

**Comment by ayrtonmassey** [comment]
If you inspect the DOM of the search box you can see that the input element causes the overflow.

See this: [Why don't flex items shrink past content size?](https://stackoverflow.com/a/36247448)

> [4.5. Automatic Minimum Size of Flex Items](https://www.w3.org/TR/css-flexbox-1/#min-size-auto)
> 
> To provide a more reasonable default minimum size for flex items, this specification introduces a new  `auto` value as the initial value of the `min-width` and `min-height` properties defined in CSS 2.1.

Edge appears to follow the spec - if you inspect the `<input>`/`.ms-SearchBox-field` element the computed value for `min-width` is `auto`. Setting `min-width` to `0px` fixes the issue.

It's possible that Edge is computing the auto-width incorrectly (why is it `181px`?) but I think the fix should be applied within Office Fabric.

**Comment by ayrtonmassey** [comment]
I had a chat with the Edge team and we confirmed that Edge is following the spec, and they do not plan to change this behaviour (so Office Fabric needs to apply the fix). Scroll down for TL;DR.

---

According to the flexbox spec:

- [§9.7. 4. d.](https://www.w3.org/TR/css-flexbox-1/#resolve-flexible-lengths) When calculating the size of a flex item the item's size is *clamped* by its `min/max-size` (where `size` is `width` or `height`, depending on the flex direction).

- [§4.5.](https://www.w3.org/TR/css-flexbox-1/#automatic-minimum-size) Flex items have an *automatic minimum size*, which is defined as:

    >... the smaller of [the element's] *content size* and its *specified size* ...

We can ignore the rest of the paragraph because this is the only line that applies in our case.

- The *specified size* is the value of the `width` or `height` property (depending on `flex-direction`) if the value of that property is definite.
- The *content size* is the minimum width required to display the element's content without overflow. 

For the `<input>` element in the `SearchBox` component:

- `width` is not set.
- `min-width` is implicitly `auto`.

`min-width: auto` triggers the *automatic minimum size* logic. There is no *specified size* because `width` is not set, so the *content size* is used. The *content size* for input elements is based on the `size` attribute of the input, which is 20 characters by default.

Thus the default `min-width` for flexible `<input>` elements with no further attributes is 20 characters. The `flex-basis` does not affect this minimum width. When the size of the `<input>` is calculated it is clamped by this `min-width` and thus will not shrink below 20 characters.

## Why has the behaviour changed in Edge Fall '18?

The Edge team has made a change to more closely align with the spec. In the current version of `Edge` a `flex-basis` of `0px` implicitly sets `width: 0px` or `height: 0px`, depending on `flex-direction`. **This behaviour has been removed in Fall '18.**

## Why does it work in Chrome?

According to [the spec](https://www.w3.org/TR/css-flexbox-1/#valdef-min-width-auto):

> On a flex item whose `overflow` is visible in the *main axis*, when specified on the flex item’s *main-axis* `min-size` property, specifies an *automatic minimum size*. It otherwise computes to 0 (unless otherwise defined by a future specification).

The `<input>` of the `SearchBox` has `overflow: hidden;` set. Chrome considers this property when computing the `min-width` and sets it to 0, as defined above. However, according to the spec `overflow` is only applicable to block containers, flexbox containers or grid containers. `<input>` is none of these so Edge ignores the `overflow` property and applies the *automatic minimum size* of 20 characters. If you remove the `overflow: hidden;` property from the `<input>` the same issue reproduces in Chrome.

## Examples

See the following CodePen. 

https://codepen.io/anon/pen/EdPQQR

- The container has `width: 40px`, `display: flex` and `flex-direction: row`.
- The `<input>` has `flex: 1 1 0px`.

The examples are as follows:

1. `input` with default value for all properties.
   - [Chrome] The element overflows and has a width of 20 characters.
   - [Edge Current] The element shrinks to fit the container.
   - [Edge Fall '18] The element overflows and has a width of 20 characters.
1. `input` with `size` set to 10.
   - [Chrome] The element overflows and has a width of 10 characters.
   - [Edge Current] The element shrinks to fit the container.
   - [Edge Fall '18] The element overflows and has a width of 10 characters.
1. `input` with `overflow: hidden;`. 
   - [Chrome] The element shrinks to fit the container.
   - [Edge Current] The element shrinks to fit the container.
   - [Edge Fall '18] The element overflows and has a width of 20 characters.
1. `input` with `min-width: 0;`.
   - [Chrome] The element shrinks to fit the container.
   - [Edge Current] The element shrinks to fit the container.
   - [Edge Fall '18] The element shrinks to fit the container.
1. `input` with `width: 0;`.
   - [Chrome] The element shrinks to fit the container.
   - [Edge Current] The element shrinks to fit the container.
   - [Edge Fall '18] The element shrinks to fit the container.


## TL;DR:

The default `min-width` for flexible `<input>` elements with no further attributes is 20 characters. The `flex-basis` does not affect this minimum width. When the size of the `<input>` is calculated it is clamped by this `min-width` and thus will not shrink below 20 characters.

To allow the `<input>` to shrink below 20 characters, define `min-width: 0px;` on the `.ms-SearchBox-field` element of the `SearchBox` component.

**Comment by micahgodbolt** [comment]
@ayrtonmassey Thanks for digging in with the Edge team. I think that was the missing piece here.....why are they changing, and was it on purpose or a regression. We'll take a look.
